tests/data_context/test_data_context_datasources.py

```diff
@@ -58,34 +59,146 @@ def test_data_context_instantiates_inline_store_backend_with_filesystem_config(
+
+    assert not mock_get.called
+
+
+@pytest.mark.parametrize(
+    "data_context_fixture_name",
+    [
+        # In order to leverage existing fixtures in parametrization, we provide
+        # their string names and dynamically retrieve them using pytest's built-in
+        # `request` fixture.
+        # Source: https://stackoverflow.com/a/64348247
+        pytest.param(
+            "in_memory_runtime_context",
+            id="BaseDataContext",
+        ),
+        pytest.param(
+            "cloud_data_context_in_cloud_mode_with_datasource_pandas_engine",
+            id="DataContext",
+        ),
+    ],
+)
+@pytest.mark.unit
+def test_get_datasource_cache_miss(
+    data_context_fixture_name: str,
+    request,
+) -> None:
+    """
+    What does this test and why?
+
+    For persistence-enabled contexts, we should leverage the underlying store in the case
+    of a cache miss.
+
+    For persistence-disabled contexts, we only look at the cache.
+    """
+    context = request.getfixturevalue(data_context_fixture_name)
+
+    name = "my_fake_datasource_name"
+
+    # Initial GET will miss the cache, necessitating store retrieval
+    with mock.patch(
+        "great_expectations.data_context.store.DatasourceStore.has_key"
+    ), mock.patch(
+        "great_expectations.data_context.data_context.BaseDataContext._instantiate_datasource_from_config"
+    ), mock.patch(
+        "great_expectations.data_context.store.DatasourceStore.get"
+    ) as mock_get:
+        context.get_datasource(name)
+
+    assert mock_get.called
+
+    # Subsequent GET will retrieve from the cache
+    with mock.patch(
+        "great_expectations.data_context.store.DatasourceStore.get"
+    ) as mock_get:
+        context.get_datasource(name)
+
+    assert not mock_get.called
+
+
+@pytest.mark.unit
+def test_DataContext_add_datasource_updates_cache_and_store(
+    cloud_data_context_in_cloud_mode_with_datasource_pandas_engine: DataContext,
+) -> None:
+    """
+    What does this test and why?
+
+    For persistence-enabled contexts, we should update both the cache and the underlying
+    store upon adding a datasource.
+    """
+    context = cloud_data_context_in_cloud_mode_with_datasource_pandas_engine
 
     name = "my_new_datasource"
 
     assert name not in context.datasources
 
-    context.add_datasource(
-        name,
-        class_name="PandasDatasource",
-        batch_kwargs_generators={
-            "subdir_reader": {
-                "class_name": "SubdirReaderBatchKwargsGenerator",
-            }
-        },
-    )
-
+    with mock.patch(
+        "great_expectations.data_context.store.DatasourceStore.set"
+    ) as mock_set:
+        context.add_datasource(
+            name,
+            class_name="PandasDatasource",
+            batch_kwargs_generators={
+                "subdir_reader": {
+                    "class_name": "SubdirReaderBatchKwargsGenerator",
+                }
+            },
+        )
```

Comment:
Maybe two tests? Label this one with the v2_api flag and another with a v3 api datasource?

Reply:
since we're moreso testing the persistence and not the actual datasource, could I just use a v3 datasource instead? Don't think we necessarily need two tests for v2 vs v3 when the underlying behavior is the same in regards to caching.

Reply:
ended up using the same v3 config for all tests - let me know if you disagree and I can come back 😄